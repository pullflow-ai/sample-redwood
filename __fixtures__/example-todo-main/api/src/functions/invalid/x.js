import { logger } from 'src/lib/logger'

export const handler = async (event, context) => {
  logger.info('Invoked x function')

  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      data: 'x function',
    }),
  }
}
