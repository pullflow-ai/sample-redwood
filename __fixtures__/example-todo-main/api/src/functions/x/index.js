export const value = 'HELLO from X'
