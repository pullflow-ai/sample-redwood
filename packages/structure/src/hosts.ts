import fg from 'fast-glob'
import * as fs from 'fs-extra'

import type { Paths } from '@redwoodjs/project-config'
import { getPaths } from '@redwoodjs/project-config'

import { lazy } from './x/decorators'

/**
 * The host interface allows us to decouple the "model/*"
 * classes from access to the file system.
 * This is critical for editor support (ex: showing diagnostics on unsaved files)
 */
export interface Host {
  existsSync(path: string): boolean
  readFileSync(path: string): string
  readdirSync(path: string): string[]
  globSync(pattern: string): string[]
  writeFileSync(path: string, contents: string): void
  paths: Paths
}

export class DefaultHost implements Host {
  existsSync(path: string) {
    return fs.existsSync(path)
  }
  readFileSync(path: string) {
    return fs.readFileSync(path, { encoding: 'utf8' }).toString()
  }
  readdirSync(path: string) {
    return fs.readdirSync(path)
  }
  globSync(pattern: string) {
    // globSync only works with / as the path separator, even on Windows
    return fg.sync(pattern.replaceAll('\\', '/'))
  }
  writeFileSync(path: string, contents: string) {
    return fs.writeFileSync(path, contents)
  }
  @lazy()
  get paths() {
    return getPaths()
  }
}
