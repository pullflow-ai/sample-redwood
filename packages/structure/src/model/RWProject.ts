import { join } from 'path'
import type { DMMF } from '@prisma/generator-helper'
import { getDMMF, getSchema } from '@prisma/internals'
import { getPaths, processPagesDir } from '@redwoodjs/project-config'
import type { Host } from '../hosts'
import { BaseNode } from '../ide'
import { lazy, memo } from '../x/decorators'
import {
  followsDirNameConvention,
  isCellFileName,
  isLayoutFileName,
} from '../x/path'
import { URL_file } from '../x/URL'
import { RWCell } from './RWCell'
import { RWComponent } from './RWComponent'
import { RWEnvHelper } from './RWEnvHelper'
import { RWFunction } from './RWFunction'
import { RWLayout } from './RWLayout'
import { RWPage } from './RWPage'
import { RWRouter } from './RWRouter'
import { RWSDL } from './RWSDL'
import { RWService } from './RWService'
import { RWTOML } from './RWTOML'

export interface RWProjectOptions {
  projectRoot: string
  host: Host
}

const allFilesGlob = '/**/*.{js,jsx,ts,tsx}'

/**
 * Represents a Redwood project.
 * This is the root node.
 */
export class RWProject extends BaseNode {
  constructor(public opts: RWProjectOptions) {
    super()
  }

  parent = undefined

  get host() {
    return this.opts.host
  }

  get projectRoot() {
    return this.opts.projectRoot
  }

  @lazy() get id() {
    return URL_file(this.projectRoot)
  }

  children() {
    return [
      this.redwoodTOML,
      ...this.pages,
      this.router,
      ...this.services,
      ...this.sdls,
      ...this.layouts,
      ...this.components,
      this.envHelper,
    ]
  }

  @lazy() get pathHelper() {
    return getPaths(this.projectRoot)
  }

  @lazy() get isTypeScriptProject(): boolean {
    return (
      this.host.existsSync(join(this.pathHelper.web.base, 'tsconfig.json')) ||
      this.host.existsSync(join(this.pathHelper.api.base, 'tsconfig.json'))
    )
  }

  @memo() async prismaDMMF(): Promise<DMMF.Document | undefined> {
    try {
      const datamodel = await getSchema(this.pathHelper.api.dbSchema)
      return await getDMMF({ datamodel })
    } catch {
      return undefined
    }
  }

  @memo() async prismaDMMFModelNames() {
    const dmmf = await this.prismaDMMF()
    return dmmf ? dmmf.datamodel.models.map((m) => m.name) : []
  }

  @lazy() get redwoodTOML(): RWTOML {
    return new RWTOML(join(this.projectRoot, 'redwood.toml'), this)
  }

  @lazy() private get processPagesDir() {
    try {
      return processPagesDir(this.pathHelper.web.pages)
    } catch {
      return []
    }
  }

  @lazy() get pages(): RWPage[] {
    return this.processPagesDir.map((p) => new RWPage(p.constName, p.path, this))
  }

  @lazy() get router() {
    return this.getRouter()
  }

  getRouter = () => {
    return new RWRouter(this.pathHelper.web.routes, this)
  }

  servicesFilePath(name: string) {
    const ext = this.isTypeScriptProject ? '.ts' : '.js'
    return join(this.pathHelper.api.services, name, name + ext)
  }

  @lazy() get defaultNotFoundPageFilePath() {
    const ext = this.isTypeScriptProject ? '.tsx' : '.jsx'
    return join(this.pathHelper.web.pages, 'NotFoundPage', 'NotFoundPage' + ext)
  }

  @lazy() get services() {
    return this.host
      .globSync(this.pathHelper.api.services + allFilesGlob)
      .filter(followsDirNameConvention)
      .map((x) => new RWService(x, this))
  }

  @lazy() get sdls() {
    return this.host
      .globSync(this.pathHelper.api.graphql + '/**/*.sdl.{js,ts}')
      .map((x) => new RWSDL(x, this))
  }

  @lazy() get layouts(): RWLayout[] {
    return this.host
      .globSync(this.pathHelper.web.layouts + allFilesGlob)
      .filter(followsDirNameConvention)
      .filter(isLayoutFileName)
      .map((x) => new RWLayout(x, this))
  }

  @lazy() get functions(): RWFunction[] {
    return this.host
      .globSync(this.pathHelper.api.functions + allFilesGlob)
      .map((x) => new RWFunction(x, this))
  }

  @lazy() get components(): RWComponent[] {
    return this.host
      .globSync(this.pathHelper.web.components + allFilesGlob)
      .map((file) => {
        if (isCellFileName(file)) {
          const possibleCell = new RWCell(file, this)
          return possibleCell.isCell ? possibleCell : new RWComponent(file, this)
        }
        return new RWComponent(file, this)
      })
  }

  @lazy() get sides() {
    return ['web', 'api']
  }

  @lazy() get mocks() {
    return this.host.globSync(this.pathHelper.web.base + '/**/*.mock.{js,ts}')
  }

  @lazy() get cells(): RWCell[] {
    return this.host
      .globSync(this.pathHelper.web.base + '/**/*Cell.{js,jsx,tsx}')
      .map((file) => new RWCell(file, this))
      .filter((file) => file.isCell)
  }

  @lazy() get envHelper(): RWEnvHelper {
    return new RWEnvHelper(this)
  }
}
