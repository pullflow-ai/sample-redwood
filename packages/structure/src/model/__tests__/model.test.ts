/* eslint-disable @typescript-eslint/no-unused-expressions */
import { basename, resolve } from 'path'

import { describe, it, expect } from 'vitest'

import { DefaultHost } from '../../hosts'
import { URL_file } from '../../x/URL'
import { RWProject } from '../RWProject'

describe('Redwood Project Model', () => {
  it('can process example-todo-main', async () => {
    const projectRoot = getFixtureDir('example-todo-main')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })

    const pageNames = new Set(project.pages.map((p) => p.basenameNoExt))
    expect(pageNames).toEqual(
      new Set([
        'FatalErrorPage',
        'HomePage',
        'NotFoundPage',
        'PrivatePage',
        'TypeScriptPage',
        'EditUserPage',
        'FooPage',
        'BarPage',
      ]),
    )
    for (const page of project.pages) {
      page.basenameNoExt
      page.route?.id
    }
    expect(
      project.sdls.map((s) => s.name).sort((a, b) => (a < b ? -1 : 1)),
    ).toEqual(['currentUser', 'todos'])

    for (const c of project.components) {
      c.basenameNoExt
    }
    project.components.length
    project.components.map((c) => c.basenameNoExt)
    project.functions.length
    project.services.length
    project.sdls.length
    const ds = await project.collectDiagnostics()
    ds.length
    const uri = URL_file(projectRoot, 'api/src/graphql/todos.sdl.js')
    const node = await project.findNode(uri)
    expect(node).toBeDefined()
    expect(node?.id).toEqual(uri)
    if (node) {
      const info = await node.collectIDEInfo()
      info.length
      info
    }
  })

  it('example-todo-main-with-errors', async () => {
    const projectRoot = getFixtureDir('example-todo-main-with-errors')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })
    const ds = await project.collectDiagnostics()
    expect(ds.length).toBeGreaterThan(0)
    // const diagnosticCodes = new Set(ds.map((d) => d.diagnostic.code))
    // expect(diagnosticCodes).toEqual(
    //   new Set([RWError.NOTFOUND_PAGE_NOT_DEFINED])
    // )
    const dss = await project.router.collectDiagnostics()
    expect(dss.length).toBeGreaterThan(0)
  })
})

describe('Cells', () => {
  it('Correctly determines a Cell component vs a normal component', () => {
    const projectRoot = getFixtureDir('example-todo-main-with-errors')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })
    const cells = project.cells
    expect(cells).toHaveLength(1)
    expect(cells.map((cell) => basename(cell.filePath))).not.toContain(
      'TableCell.js',
    )
  })

  it('Can get the operation name of the QUERY', () => {
    const projectRoot = getFixtureDir('example-todo-main')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })
    const cell = project.cells.find((x) => x.uri.endsWith('TodoListCell.tsx'))
    expect(cell?.queryOperationName).toMatch('TodoListCell_GetTodos')
  })

  it('Warns you when you do not supply a name to QUERY', async () => {
    const projectRoot = getFixtureDir('example-todo-main-with-errors')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })

    const cell = project.cells.find((x) => x.uri.endsWith('TodoListCell.js'))
    const x = await cell?.collectDiagnostics()
    expect(x).not.toBeUndefined()
    expect(x?.map((e) => e.diagnostic.message)).toContain(
      'We recommend that you name your query operation',
    )
  })
})

describe('Redwood Page detection', () => {
  it('detects pages', () => {
    const projectRoot = getFixtureDir('example-todo-main')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })
    const routes = project.getRouter().routes
    const pages = routes.map((r) => r.page).sort()
    const pageConstants = pages.map((p) => p?.constName)
    // Note: Pages can be duplicated if used by multiple routes, we use a Set
    expect(pageConstants).toEqual([
      'HomePage',
      'TypeScriptPage',
      'FooPage',
      'BarPage',
      'PrivatePage',
      'PrivatePage',
      'PrivatePage',
      'NotFoundPage',
      undefined,
    ])
  })
})

describe('Redwood Route detection', () => {
  it('detects the page identifier for a route', () => {
    const projectRoot = getFixtureDir('example-todo-main')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })
    const routes = project.getRouter().routes

    const pageIdentifiers = routes.map((r) => r.page_identifier_str)

    expect(pageIdentifiers.length).toBe(9)
    expect(pageIdentifiers).toMatchSnapshot()
  })
  it('detects routes with the prerender prop', () => {
    const projectRoot = getFixtureDir('example-todo-main')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })
    const routes = project.getRouter().routes

    const prerenderRoutes = routes
      .filter((r) => r.prerender)
      // Make it a little easier to read by only keeping the attributes we're
      // interested in
      .map(({ name, path }) => ({ name, path }))

    expect(prerenderRoutes.length).toBe(8)
    expect(prerenderRoutes).toContainEqual({ name: 'home', path: '/' })
    expect(prerenderRoutes).toContainEqual({
      name: 'typescriptPage',
      path: '/typescript',
    })
    expect(prerenderRoutes).toContainEqual({
      name: 'someOtherPage',
      path: '/somewhereElse',
    })
    expect(prerenderRoutes).toContainEqual({ name: 'fooPage', path: '/foo' })
    expect(prerenderRoutes).toContainEqual({ name: 'barPage', path: '/bar' })
    expect(prerenderRoutes).toContainEqual({
      name: 'privatePage',
      path: '/private-page',
    })
  })
  it('detects authenticated routes', () => {
    const projectRoot = getFixtureDir('example-todo-main')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })
    const routes = project.getRouter().routes

    const authenticatedRoutes = routes
      .filter((r) => r.isPrivate)
      .map(({ name, path, unauthenticated, roles }) => ({
        name,
        path,
        unauthenticated,
        roles,
      }))

    expect(authenticatedRoutes.length).toBe(3)
  })

  it('detects name and path for an authenticated route', () => {
    const projectRoot = getFixtureDir('example-todo-main')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })
    const routes = project.getRouter().routes

    const authenticatedRoutes = routes
      .filter((r) => r.isPrivate)
      .map(({ name, path, unauthenticated, roles }) => ({
        name,
        path,
        unauthenticated,
        roles,
      }))

    expect(authenticatedRoutes[1].name).toBe('privatePageAdmin')
    expect(authenticatedRoutes[1].path).toBe('/private-page-admin')
    expect(authenticatedRoutes[1].unauthenticated).toBe('home')
    expect(authenticatedRoutes[1].roles).toBeTypeOf('string')
    expect(authenticatedRoutes[1].roles).toContain('admin')
  })

  it('detects roles for an authenticated route when roles is a string of a single role', () => {
    const projectRoot = getFixtureDir('example-todo-main')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })
    const routes = project.getRouter().routes

    const authenticatedRoutes = routes
      .filter((r) => r.isPrivate)
      .map(({ name, path, unauthenticated, roles }) => ({
        name,
        path,
        unauthenticated,
        roles,
      }))

    expect(authenticatedRoutes[1].name).toBe('privatePageAdmin')
    expect(authenticatedRoutes[1].path).toBe('/private-page-admin')
    expect(authenticatedRoutes[1].unauthenticated).toBe('home')
    expect(authenticatedRoutes[1].roles).toBeTypeOf('string')
    expect(authenticatedRoutes[1].roles).toContain('admin')
  })

  it('detects roles for an authenticated route when roles is an array of a roles', () => {
    const projectRoot = getFixtureDir('example-todo-main')
    const project = new RWProject({ projectRoot, host: new DefaultHost() })
    const routes = project.getRouter().routes

    const authenticatedRoutes = routes
      .filter((r) => r.isPrivate)
      .map(({ name, path, unauthenticated, roles }) => ({
        name,
        path,
        unauthenticated,
        roles,
      }))

    expect(authenticatedRoutes[2].name).toBe('privatePageAdminSuper')
    expect(authenticatedRoutes[2].path).toBe('/private-page-admin-super')
    expect(authenticatedRoutes[2].unauthenticated).toBe('home')
    expect(authenticatedRoutes[2].roles).toBeInstanceOf(Array)
    expect(authenticatedRoutes[2].roles).toContain('owner')
    expect(authenticatedRoutes[2].roles).toContain('superuser')
    expect(authenticatedRoutes[2].roles).not.toContain('member')
  })
})

function getFixtureDir(
  name:
    | 'example-todo-main-with-errors'
    | 'example-todo-main'
    | 'empty-project'
    | 'test-project',
) {
  return resolve(__dirname, `../../../../../__fixtures__/${name}`)
}
