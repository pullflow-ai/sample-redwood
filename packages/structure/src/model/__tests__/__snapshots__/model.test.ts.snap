// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Redwood Route detection > detects the page identifier for a route 1`] = `
[
  "HomePage",
  "TypeScriptPage",
  "EditUserPage",
  "FooPage",
  "BarPage",
  "PrivatePage",
  "PrivatePage",
  "PrivatePage",
  "NotFoundPage",
]
`;
